src/TestFramework/Microsoft.Rest.ClientRuntime.Azure.TestFramework/ConnectionString.cs

```diff
@@ -0,0 +1,308 @@
+        /// </summary>
+        private void NormalizeKeyValuePairs()
+        {
+            string clientId, spn, password, spnSecret, userId;
+            KeyValuePairs.TryGetValue(ConnectionStringKeys.AADClientIdKey.ToLower(), out clientId);
+            KeyValuePairs.TryGetValue(ConnectionStringKeys.ServicePrincipalKey.ToLower(), out spn);
+
+            KeyValuePairs.TryGetValue(ConnectionStringKeys.UserIdKey.ToLower(), out userId);
+            KeyValuePairs.TryGetValue(ConnectionStringKeys.PasswordKey.ToLower(), out password);
+            KeyValuePairs.TryGetValue(ConnectionStringKeys.ServicePrincipalSecretKey.ToLower(), out spnSecret);            
+
+            //ClientId was provided and servicePrincipal was empty, we want ServicePrincipal to be initialized
+            //At some point we will deprecate ClientId keyName
+            if (!string.IsNullOrEmpty(clientId) && (string.IsNullOrEmpty(spn)))
+            {
+                KeyValuePairs[ConnectionStringKeys.ServicePrincipalKey.ToLower()] = clientId;
+            }
+
+            //Set the value of PasswordKey to ServicePrincipalSecret ONLY if userId is empty
+            //If UserId is not empty, we are not sure if it's a password for inter active login or ServicePrincipal SecretKey
+            if (!string.IsNullOrEmpty(password) && (string.IsNullOrEmpty(spnSecret)) && (string.IsNullOrEmpty(userId)))
+            {
+                KeyValuePairs[ConnectionStringKeys.ServicePrincipalSecretKey.ToLower()] = password;
+            }
+
+            //Initialize default values if found empty
+            if (string.IsNullOrEmpty(clientId) && (string.IsNullOrEmpty(spn)))
+            {
+                KeyValuePairs.UpdateDictionary(ConnectionStringKeys.ServicePrincipalKey, "1950a258-227b-4e31-a9cf-717495945fc2");
+            }
+
+            //Initialize raw tokens to a non-null/non-empty strings
+            KeyValuePairs.UpdateDictionary(ConnectionStringKeys.RawTokenKey, ConnectionStringKeys.RawTokenKey);
+            KeyValuePairs.UpdateDictionary(ConnectionStringKeys.RawGraphTokenKey, ConnectionStringKeys.RawGraphTokenKey);
+        }
+
+        /// <summary>
+        /// Detect any connection string violations
+        /// </summary>
+        private void DetectViolations()
+        {
+            //So far only 1 violation is being checked
+            //We should also check if a supported Environment is provided in connection string (but seems we do not throw exception for unsupported environment)
+            bool envSet = IsEnvironmentSet();
+            string nonEmptyUriKey = FirstNonNullUriInConnectionString();
+
+            if(envSet)
+            {
+                if(!string.IsNullOrEmpty(nonEmptyUriKey))
+                {
+                    string envName = KeyValuePairs.GetValueUsingCaseInsensitiveKey(ConnectionStringKeys.EnvironmentKey);
```

Comment:
Why?  Why don't we just set the endpoints based on the envrionment and override with other settings?  I know this is how it worked before, but I can't see any functional value this adds.